File: services/mini-coupang/src/test/java/com/deepblog/minicoupang/domain/product/storage/ProductRepositoryIntegrationTest.java
package com.deepblog.minicoupang.domain.product.storage;

import com.deepblog.minicoupang.domain.product.Product;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.testcontainers.service.connection.ServiceConnection;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@SpringBootTest
@Testcontainers
class ProductRepositoryIntegrationTest {

    @Container
    @ServiceConnection
    static PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>("postgres:16");

    @Autowired
    ProductRepository repository;

    @Test
    void save_and_find_product_round_trip() {
        Product saved = repository.save(
                Product.create(1L, "Coupang Seller", "스마트폰", 1_200_000L, 10));

        assertThat(saved.getId()).isNotNull();

        Product found = repository.findById(saved.getId()).orElseThrow();
        assertThat(found.getName()).isEqualTo("스마트폰");
        assertThat(found.getPrice()).isEqualTo(1_200_000L);
        assertThat(found.getStock()).isEqualTo(10);
        assertThat(found.getSellerId()).isEqualTo(1L);
        assertThat(found.getSellerName()).isEqualTo("Coupang Seller");
    }

    @Test
    void findBySellerId_returns_products_for_that_seller_only() {
        repository.saveAll(List.of(
                Product.create(42L, "S42", "A", 100L, 1),
                Product.create(42L, "S42", "B", 200L, 2),
                Product.create(99L, "S99", "C", 300L, 3)
        ));

        List<Product> s42Products = repository.findBySellerId(42L);
        assertThat(s42Products).hasSize(2);
        assertThat(s42Products).extracting(Product::getName)
                .containsExactlyInAnyOrder("A", "B");
    }

    @Test
    void negative_price_is_rejected_by_factory() {
        assertThatThrownBy(() -> Product.create(1L, "S", "X", -1L, 0))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void decreaseStock_reduces_stock() {
        Product product = Product.create(1L, "S", "X", 100L, 5);
        product.decreaseStock(2);
        assertThat(product.getStock()).isEqualTo(3);
    }

    @Test
    void decreaseStock_throws_when_insufficient() {
        Product product = Product.create(1L, "S", "X", 100L, 1);
        assertThatThrownBy(() -> product.decreaseStock(2))
                .isInstanceOf(IllegalStateException.class);
    }
}

File: services/product-service/src/test/kotlin/com/deepblog/product/storage/ProductRepositoryIntegrationTest.kt
package com.deepblog.product.storage

import com.deepblog.product.domain.Product
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.testcontainers.service.connection.ServiceConnection
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers

@SpringBootTest
@Testcontainers
class ProductRepositoryIntegrationTest {

    companion object {
        @Container
        @ServiceConnection
        @JvmStatic
        val postgres: PostgreSQLContainer<*> = PostgreSQLContainer("postgres:16")
    }

    @Autowired lateinit var repository: ProductRepository

    @Test
    fun `save and find product round-trip`() {
        val saved = repository.save(
            Product(
                sellerId = 1L,
                sellerName = "Coupang Seller",
                name = "스마트폰",
                price = 1_200_000L,
                stock = 10,
            ),
        )

        assertThat(saved.id).isNotNull()

        val found = repository.findById(saved.id!!).orElseThrow()
        assertThat(found.name).isEqualTo("스마트폰")
        assertThat(found.price).isEqualTo(1_200_000L)
        assertThat(found.stock).isEqualTo(10)
        assertThat(found.sellerId).isEqualTo(1L)
        assertThat(found.sellerName).isEqualTo("Coupang Seller")
    }

    @Test
    fun `findBySellerId returns products for that seller only`() {
        repository.saveAll(
            listOf(
                Product(sellerId = 42L, sellerName = "S42", name = "A", price = 100, stock = 1),
                Product(sellerId = 42L, sellerName = "S42", name = "B", price = 200, stock = 2),
                Product(sellerId = 99L, sellerName = "S99", name = "C", price = 300, stock = 3),
            ),
        )

        val s42Products = repository.findBySellerId(42L)
        assertThat(s42Products).hasSize(2)
        assertThat(s42Products.map { it.name }).containsExactlyInAnyOrder("A", "B")
    }

    @Test
    fun `negative price is rejected at entity init`() {
        val ex = runCatching {
            Product(sellerId = 1L, sellerName = "S", name = "X", price = -1, stock = 0)
        }.exceptionOrNull()
        assertThat(ex).isInstanceOf(IllegalArgumentException::class.java)
    }
}
